### ISSUE TITLE
In multi-line mode, ^ matches after a trailing newline

### ISSUE THREAD

**Issue by robinst** [rust-lang/regex#406]
In trishume/syntect#34, we have the pattern `r"(?m)^\s*$"`. The input string `"test\n"` unexpectedly matches that pattern:

```rust
let re = Regex::new(r"(?m)^\s*$").unwrap();
assert!(!re.is_match("test\n"));
// thread 'main' panicked at 'assertion failed: !re.is_match("test\n")'
```

The regex crate behaves differently here than e.g. Oniguruma (which doesn't match). We can create a smaller case which I think is caused by the same problem:

```rust
let re = Regex::new(r"(?m)^").unwrap();
assert_eq!(re.find_iter("test\n").count(), 1);
// thread 'main' panicked at 'assertion failed: `(left == right)` left: `2`, right: `1`'
```

Instead of matching once, it matches twice, once for position 0 and another time for position 5 (after `\n`).

To me it looks like regex currently implements the concept of `\n` being a line separator (between lines), instead of a line terminator (at the end of a line).

Is this intentional behavior? If no, do you think it would be hard to change? I can look into it, but some pointers would be appreciated :).

FWIW, Rust itself doesn't think that `"test\n"` is two lines:

```rust
let lines: Vec<_> = "test\n".lines().collect();
println!("{:?}", lines);
// ["test"]
```

**Comment by BurntSushi** [comment]
Unfortunately yes this is correct. It is in fact consistent with RE2-style of matching. From [Russ's article](https://swtch.com/~rsc/regexp/regexp3.html):

> In single-line mode, if the input text ends with a newline character, Perl and PCRE allow $ to match either before or after that final newline. RE2 requires that it match after, at the very end of the text.
>
> Similarly, in multi-line mode, if the input text ends with a newline character, Perl and PCRE do not allow ^, which normally matches following a newline, to match at the very end of the text. RE2 does.

You can see that behavior is consistent with Go's regexp engine (which is also RE2-style): https://play.golang.org/p/Zc-J0VPpob

Interestingly, Python (which has a backtracking engine) agrees as well:

```
>>> import re
>>> x = re.compile('(?m)^')
>>> [g.span() for g in x.finditer('test\n')]
[(0, 0), (5, 5)]
```

Given all of this, I'm inclined to keep current behavior. What kind of work-arounds are available to you?

**Comment by robinst** [comment]
Thanks for the links! Do we know *why* RE2 behaves that way?

For Python, it's noteworthy that `splitlines()` doesn't consider it to be two lines either:

```python
>>> 'test\n'.splitlines()
['test']
```

> Given all of this, I'm inclined to keep current behavior. (And in my mind, the current behavior seems more consistent.)

It's interesting that it makes it inconsistent with what Rust's `lines` method does. But it looks like Python has the same inconsistency.

Perl seems to behave the same way as Oniguruma (my Perl is rusty, but `scalar` makes it print the size of the array):

```
$ perl -e 'my @matches = "test\n" =~ /(?m)^/g; print scalar @matches; print "\n"'
1
```

Seems like this is another one of those things that is inconsistent between engines and should not be relied on. Either way, we should document the behavior. I'll raise another PR once #405 is merged.

**Comment by robinst** [comment]
Oh, I forgot to mention a workaround: I think we can rewrite the regex to use `\A` instead of `^`, because we only match with individual lines anyway.

That means we have to do some parsing because we don't want to replace it in e.g. character classes. But we have to do that for other reasons already.

**Comment by BurntSushi** [comment]
I don't know specifically why RE2 behaves this way, but I'm pretty sure that if we wanted the exact PCRE behavior, we would need to special case it. The absence of the special case results in the behavior we have today.

**Comment by robinst** [comment]
That sounds reasonable. I imagine the special case would mean looking ahead to see if we're at the end of the string or not, which is a bit weird for `^`.
